Index: /arch/mips/Kconfig
===================================================================
--- /arch/mips/Kconfig
+++ /arch/mips/Kconfig
@@ -395,6 +395,7 @@ config LANTIQ
 	select GPIOLIB
 	select SWAP_IO_SPACE
 	select BOOT_RAW
+	select COMMON_CLK
 	select CLKDEV_LOOKUP
 	select USE_OF
 	select PINCTRL

Index: /drivers/clk/Makefile
===================================================================
--- /drivers/clk/Makefile
+++ /drivers/clk/Makefile
@@ -66,6 +66,7 @@ obj-y					+= imgtec/
 obj-$(CONFIG_ARCH_MXC)			+= imx/
 obj-$(CONFIG_MACH_INGENIC)		+= ingenic/
 obj-$(CONFIG_ARCH_KEYSTONE)		+= keystone/
+obj-$(CONFIG_SOC_XWAY)			+= lantiq/
 obj-$(CONFIG_MACH_LOONGSON32)		+= loongson1/
 obj-$(CONFIG_ARCH_MEDIATEK)		+= mediatek/
 obj-$(CONFIG_COMMON_CLK_AMLOGIC)	+= meson/
Index: /drivers/clk/lantiq/Makefile
===================================================================
--- /dev/null
+++ /drivers/clk/lantiq/Makefile
@@ -0,0 +1,4 @@
+obj-y				+= 
+obj-$(CONFIG_SOC_XWAY)		+= clk-xway-pmu.o
+obj-$(CONFIG_SOC_XWAY)		+= clk-xway-pll.o
+obj-$(CONFIG_SOC_XWAY)		+= clk-xway-mmio-mux.o


